# ML-10 — Content Action Playbook

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/flyrank-bih/flyrank-ml-internship-starter/blob/main/work/notebooks/w07_action_playbook.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Ranked actions + reason codes

*The queue: what to do first, and why, in words a human trusts.*

In [9]:
# ============================================================
# SETUP — reconnect + rebuild features/label/split (same as w06)
# ============================================================
%pip -q install duckdb huggingface_hub scikit-learn pandas numpy

import os, getpass
import duckdb, pandas as pd, numpy as np
from sklearn.model_selection import GroupShuffleSplit
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score

HF_TOKEN = os.environ.get('HF_TOKEN')
if not HF_TOKEN:
    try:
        from google.colab import userdata
        HF_TOKEN = userdata.get('HF_TOKEN')
    except Exception:
        pass
HF_TOKEN = HF_TOKEN or getpass.getpass('HF token: ')

con = duckdb.connect()
con.sql("INSTALL httpfs; LOAD httpfs;")
con.sql(f"CREATE OR REPLACE SECRET hf (TYPE huggingface, TOKEN '{HF_TOKEN}')")

REL = "hf://datasets/FlyRank/internship-warehouse"
feature_paths = [f"{REL}/fact_content_daily_performance/month={m}/data_0.parquet"
                  for m in ["2026-01", "2026-02", "2026-03"]]
label_path = f"{REL}/fact_content_daily_performance/month=2026-04/data_0.parquet"

features = con.sql(f"""
    SELECT content_hash_id, client_hash_id,
        SUM(gsc_impressions) AS impressions_janmar,
        SUM(gsc_clicks) AS clicks_janmar,
        AVG(gsc_avg_position) FILTER (WHERE gsc_avg_position > 0) AS avg_position_janmar,
        SUM(ga4_sessions) AS sessions_janmar,
        SUM(scroll_events) AS scroll_events_janmar,
        SUM(CASE WHEN report_date >= DATE '2026-03-01' THEN gsc_impressions ELSE 0 END) AS impressions_march
    FROM read_parquet({feature_paths})
    WHERE gsc_data_available IS TRUE
    GROUP BY content_hash_id, client_hash_id
    HAVING impressions_march >= 50
""").df()

april = con.sql(f"""
    SELECT content_hash_id, client_hash_id, SUM(gsc_impressions) AS impressions_april
    FROM '{label_path}' WHERE gsc_data_available IS TRUE
    GROUP BY content_hash_id, client_hash_id
""").df()
data = features.merge(april, on=["content_hash_id","client_hash_id"], how="left")
data["impressions_april"] = data["impressions_april"].fillna(0)
data["is_declining_label"] = (data["impressions_april"] < 0.8 * data["impressions_march"]).astype(int)

def engineer(df_, median_pos):
    df_ = df_.copy()
    df_["has_position_data"] = (df_["avg_position_janmar"] > 0).astype(int)
    df_["avg_position_janmar"] = df_["avg_position_janmar"].replace(0, median_pos)
    for col in ["impressions_janmar","clicks_janmar","sessions_janmar","scroll_events_janmar","impressions_march"]:
        df_[f"log_{col}"] = np.log1p(df_[col])
    return df_

FEATURES_V2 = ["log_impressions_janmar","log_clicks_janmar","avg_position_janmar",
               "has_position_data","log_sessions_janmar","log_scroll_events_janmar"]  # WITHOUT log_impressions_march, per w06's leakage finding

gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
train_idx, test_idx = next(gss.split(data, data["is_declining_label"], groups=data["client_hash_id"]))
train_grp = data.iloc[train_idx].reset_index(drop=True)
test_grp = data.iloc[test_idx].reset_index(drop=True)

median_pos = train_grp.loc[train_grp["avg_position_janmar"] > 0, "avg_position_janmar"].median()
train_grp = engineer(train_grp, median_pos)
test_grp = engineer(test_grp, median_pos)

rf = RandomForestClassifier(n_estimators=200, max_depth=10, min_samples_leaf=25,
                             class_weight="balanced_subsample", random_state=42, n_jobs=-1)
rf.fit(train_grp[FEATURES_V2], train_grp["is_declining_label"])
scores_grp = rf.predict_proba(test_grp[FEATURES_V2])[:, 1]

print(f"Rows: {len(data):,} | Base rate: {data['is_declining_label'].mean():.3f}")
print(f"Test clients: {test_grp['client_hash_id'].nunique()} | Test AUC: {roc_auc_score(test_grp['is_declining_label'], scores_grp):.3f}")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Rows: 116,114 | Base rate: 0.518
Test clients: 9 | Test AUC: 0.554


In [20]:

test_grp['rf_score'] = scores_grp

def reason_codes(row):
    codes = []
    if row['rf_score'] >= 0.65: codes.append("model_decline_risk")
    if row['impressions_march'] >= 500: codes.append("high_visibility")
    if row['avg_position_janmar'] > 20: codes.append("weak_position")
    return "|".join(codes) if codes else "general_review"

def suggested_action(row):
    if row['rf_score'] >= 0.65 and row['impressions_march'] >= 500: return "refresh_high_priority"
    if row['rf_score'] >= 0.5: return "refresh"
    return "monitor"

test_grp['reason_codes'] = test_grp.apply(reason_codes, axis=1)
test_grp['suggested_action'] = test_grp.apply(suggested_action, axis=1)
test_grp = test_grp.sort_values('rf_score', ascending=False).reset_index(drop=True)
test_grp['rank'] = test_grp.index + 1

print(f"Clients in top 50: {test_grp.head(50)['client_hash_id'].nunique()} of {test_grp['client_hash_id'].nunique()}")
test_grp.head(20)[['rank','client_hash_id','rf_score','suggested_action','reason_codes']]
data = data.sort_values(['client_hash_id', 'content_hash_id']).reset_index(drop=True)

Clients in top 50: 5 of 9


In [21]:
import os, json
os.makedirs("work/outputs", exist_ok=True)

export_cols = ['rank','content_hash_id','client_hash_id','rf_score',
               'suggested_action','reason_codes','impressions_march']
test_grp[export_cols].to_csv("work/outputs/action_playbook_queue.csv", index=False)

metrics = {
    "label": "Jan-Mar features -> April decline (impressions_april < 0.8*impressions_march)",
    "features_used": FEATURES_V2,
    "validated_split": "client_grouped_holdout",
    "roc_auc": float(roc_auc_score(test_grp['is_declining_label'], scores_grp)),
    "precision_at_50": float(pd.Series(scores_grp).rank(ascending=False).le(50).astype(int).mul(test_grp['is_declining_label'].values).sum() / 50),
    "base_rate": float(data['is_declining_label'].mean()),
    "rows_scored": int(len(test_grp)),
    "clients_in_top50": int(test_grp.head(50)['client_hash_id'].nunique()),
}
with open("work/outputs/action_playbook_metrics.json", "w") as f:
    json.dump(metrics, f, indent=2)

print(json.dumps(metrics, indent=2))

{
  "label": "Jan-Mar features -> April decline (impressions_april < 0.8*impressions_march)",
  "features_used": [
    "log_impressions_janmar",
    "log_clicks_janmar",
    "avg_position_janmar",
    "has_position_data",
    "log_sessions_janmar",
    "log_scroll_events_janmar"
  ],
  "validated_split": "client_grouped_holdout",
  "roc_auc": 0.504344999671679,
  "precision_at_50": 0.5,
  "base_rate": 0.518421551234132,
  "rows_scored": 9089,
  "clients_in_top50": 5
}


## 2. Intended use and limits
This queue ranks pages using a Jan–Mar 2026 features → April 2026 decline model (client-grouped holdout, ROC AUC = 0.504, Precision@50 = 0.50 vs base rate 0.518). After fixing a row-ordering bug that was silently making earlier runs non-reproducible (see w05/w06), the honest, stable result shows the model performs no better than random guessing at this task with these features.

What this means in practice: this queue should NOT be used to prioritize review over random selection. It is a negative result, not a working decision-support tool — included here to document the full honest workflow, not as a production recommendation. The staleness-based baseline rule from Week 4 remains the more defensible (if still weak) ranking signal available in this project.


In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Human review + the no-go list

*What a person must check before acting. What should never be automated.*
Given the model shows no real skill, the no-go list is simple: do not use rf_score for any prioritization decision. Any reviewer using this project's output should rely on the Week-4 staleness baseline (also weak, but at least transparent and debuggable) rather than this model, until the feature set is improved and re-validated.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Monitoring / retrain triggers

*What would tell you the recommendations went stale?*
Not applicable in deployment terms, since this model is not recommended for use. For future iteration: re-attempt with month-over-month trend features (e.g. (feb_impressions - jan_impressions) / jan_impressions) before considering this label/approach viable.

In [14]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Retrain if: base rate on new data shifts >10pp from 0.534")
print("Retrain if: Precision@50 on a fresh holdout drops below ~0.60")
print("Re-audit leakage whenever a new feature is added (per w06 Section 3 method)")


Retrain if: base rate on new data shifts >10pp from 0.534
Retrain if: Precision@50 on a fresh holdout drops below ~0.60
Re-audit leakage whenever a new feature is added (per w06 Section 3 method)


## 5. Exports for the paper

*Write the queue (and any figures you want to reuse) to work/outputs/ — your paper builds on these files.*

In [17]:
import os, json
os.makedirs("work/outputs", exist_ok=True)

export_cols = ['rank','content_hash_id','client_hash_id','rf_score',
               'suggested_action','reason_codes','impressions_march']
test_grp[export_cols].to_csv("work/outputs/action_playbook_queue.csv", index=False)

metrics = {
    "label": "Jan-Mar features -> April decline (impressions_april < 0.8*impressions_march)",
    "features_used": FEATURES_V2,
    "validated_split": "client_grouped_holdout",
    "roc_auc": float(roc_auc_score(test_grp['is_declining_label'], scores_grp)),
    "precision_at_50": float(pd.Series(scores_grp).rank(ascending=False).le(50).astype(int).mul(test_grp['is_declining_label'].values).sum() / 50),
    "base_rate": float(data['is_declining_label'].mean()),
    "rows_scored": int(len(test_grp)),
    "clients_in_top50": int(test_grp.head(50)['client_hash_id'].nunique()),
}
with open("work/outputs/action_playbook_metrics.json", "w") as f:
    json.dump(metrics, f, indent=2)

print(json.dumps(metrics, indent=2))

{
  "label": "Jan-Mar features -> April decline (impressions_april < 0.8*impressions_march)",
  "features_used": [
    "log_impressions_janmar",
    "log_clicks_janmar",
    "avg_position_janmar",
    "has_position_data",
    "log_sessions_janmar",
    "log_scroll_events_janmar"
  ],
  "validated_split": "client_grouped_holdout",
  "roc_auc": 0.4989740577796261,
  "precision_at_50": 0.52,
  "base_rate": 0.518421551234132,
  "rows_scored": 9089,
  "clients_in_top50": 6
}


In [18]:
print(test_grp['suggested_action'].value_counts())
print(test_grp['reason_codes'].value_counts().head(10))

suggested_action
monitor                  5585
refresh                  3317
refresh_high_priority     187
Name: count, dtype: int64
reason_codes
general_review                                      3645
high_visibility                                     2382
weak_position                                       2189
model_decline_risk                                   268
high_visibility|weak_position                        227
model_decline_risk|weak_position                     191
model_decline_risk|high_visibility                   158
model_decline_risk|high_visibility|weak_position      29
Name: count, dtype: int64


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.